056-zk-kraft-migration.md

```diff
@@ -0,0 +1,451 @@
+If the user has an already deployed Apache Kafka cluster using a ZooKeeper ensemble for storing metadata, the Strimzi Cluster Operator has no support to migrate the running cluster to use a pool of KRaft controllers replacing the ZooKeeper nodes for the same purpose.
+
+There is also no support to upgrade a KRaft-based Apache Kafka cluster to a newer version, but this missing piece is out of the scope for this proposal.
+
+Furthermore, if the user creates one or more `KafkaNodePool` custom resources to deploy KRaft controllers, the operator itself just creates the controller nodes and restarts the Kafka brokers to use them. ZooKeeper is not used anymore but no proper migration takes place (from ZooKeeper to KRaft metadata). Because of KRaft and related migration from ZooKeeper not supported yet, this leads to the following [issue](https://github.com/strimzi/strimzi-kafka-operator/issues/9108).
+
+The ZooKeeper to KRaft migration implementation is defined by [KIP-866](https://cwiki.apache.org/confluence/display/KAFKA/KIP-866+ZooKeeper+to+KRaft+Migration).
+The upstream Apache Kafka community documentation also provides a [ZooKeeper to KRaft Migration](https://kafka.apache.org/documentation/#kraft_zk_migration) guide showing how to run the procedure manually on your cluster not deployed via the Strimzi Cluster Operator.
+
+One of the main points to highlight is that such a process doesn't support migration to a KRaft-based cluster running in "combined" mode:
+
+* Cannot migrate from a ZooKeeper based configuration to a cluster with nodes hosting both controller and broker (so actually reducing the number of nodes by removing the ZooKeeper ones).
+* The migration works by having ZooKeeper nodes to be replaced by new KRaft controller nodes.
+* Supporting "combined" mode is actually one of the rejected [alternatives](https://cwiki.apache.org/confluence/display/KAFKA/KIP-866+ZooKeeper+to+KRaft+Migration#KIP866ZooKeepertoKRaftMigration-CombinedModeMigrationSupport) in KIP-866.
+
+It means that the user environment should be sized to host more nodes running in parallel, because during the migration process both ZooKeeper nodes and KRaft controllers will be running.
+The end of the migration will involve shutting down the ZooKeeper nodes reducing the resources consumption.
+
+## Motivation
+
+The KRaft mode has been declared to be production-ready since Apache Kafka 3.3 release by the Apache Kafka project. 
+Anyway, there are still missing features as reported in the official [documentation](https://kafka.apache.org/documentation/#kraft_missing).
+The ZooKeeper mode has been deprecated after the Apache Kafka 3.5 version was released while providing the migration process to KRaft mode as preview.
+The coming Apache Kafka 3.6 release will have such a migration process supported as GA and the ZooKeeper mode will be supported until Apache Kafka 3.7 (January 2024) but removed after that.
```

Comment:
Is the 3.7 as the last 3.x release really set somewhere in Apache Kafka? Or is that some year old plan?

Reply:
I am following this plan which I linked in the proposal as well. https://cwiki.apache.org/confluence/display/KAFKA/KIP-833:+Mark+KRaft+as+Production+Ready
If it's old, stale and not used anymore I don't know. That could be more a question for Kafka committers but having  a plan in the proposal sounds to be good to me.

Reply:
I'm not sure KIP from May 2022 and modified at will afterward is up to date.

Reply:
This is the only source I have unless pinging the Kafka community and ask if it's still the plan and if not the KIP should be made obsolete and a new one should be opened with new plan. Not sure how it works there.
@tombentley @mimaison @showuon any thoughts?

Reply:
I guess you can leave the specifics out unless relying on unreliable sources. That was really my intent here.

Reply:
Where anyone has doubt about whether a KIP is out of date wrt the current state of Kafka it would help the Kafka community to ask on the users@ or dev@ mailing lists. That way the committers involved most closely with the actual implementation are able to comment authoritatively and the answer is of the widest benefit to other users contributors. We can follow up with JIRAs to improve the Kafka documentation from there. 

While the KIP might be stale, I think it's helpful to the reader of this Proposal to provide it as context.

Reply:
I took a look to the page history of KIP-833 and it seems that the latest update related to dates was on April 2023 (this is the [diff](https://cwiki.apache.org/confluence/pages/diffpagesbyversion.action?pageId=211882739&selectedPageVersions=17&selectedPageVersions=14) you get).
So it's status is not from May 2022 anymore. Of course, we are in September and we could ask to the community if this 5 months old plan is still valid otherwise the KIP should be updated again.
Regarding this proposal, I am leaving the plan as it's described but taken the dates out as suggested.